File: day01/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"regexp"
	"slices"
	"strconv"
)

const (
	puzzleInput = "input.txt"
)

var (
	digitOnlyRegex = regexp.MustCompile(`\d`)
	digitRegex     = regexp.MustCompile(`\d|one|two|three|four|five|six|seven|eight|nine`)
	revDigitRegex  = regexp.MustCompile(`\d|enin|thgie|neves|xis|evif|ruof|eerht|owt|eno`)
	words          = map[string]int{
		"one":   1,
		"two":   2,
		"three": 3,
		"four":  4,
		"five":  5,
		"six":   6,
		"seven": 7,
		"eight": 8,
		"nine":  9,
		"enin":  9,
		"thgie": 8,
		"neves": 7,
		"xis":   6,
		"evif":  5,
		"ruof":  4,
		"eerht": 3,
		"owt":   2,
		"eno":   1,
	}
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	sum1 := 0
	sum2 := 0

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		s := scanner.Bytes()
		match := digitOnlyRegex.Find(s)
		if len(match) == 0 {
			log.Fatalln("not enough digits")
		}
		firstDigit, err := parseValue(string(match))
		if err != nil {
			log.Fatalln(err)
		}
		match = digitRegex.Find(s)
		if len(match) == 0 {
			log.Fatalln("not enough digits")
		}
		first, err := parseValue(string(match))
		if err != nil {
			log.Fatalln(err)
		}
		slices.Reverse(s)
		match = digitOnlyRegex.Find(s)
		if len(match) == 0 {
			log.Fatalln("not enough digits")
		}
		lastDigit, err := parseValue(string(match))
		if err != nil {
			log.Fatalln(err)
		}
		match = revDigitRegex.Find(s)
		if len(match) == 0 {
			log.Fatalln("not enough reverse digits")
		}
		last, err := parseValue(string(match))
		if err != nil {
			log.Fatalln(err)
		}
		sum1 += firstDigit*10 + lastDigit
		sum2 += first*10 + last
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	fmt.Println("Part 1:", sum1)
	fmt.Println("Part 2:", sum2)
}

func parseValue(s string) (int, error) {
	if v, ok := words[s]; ok {
		return v, nil
	}
	return strconv.Atoi(s)
}

File: day01/src/main.rs
use once_cell::sync::Lazy;
use regex::Regex;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

static DIGIT_ONLY_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"\d").unwrap());
static DIGIT_REGEX: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"\d|one|two|three|four|five|six|seven|eight|nine").unwrap());
static REV_DIGIT_REGEX: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"\d|enin|thgie|neves|xis|evif|ruof|eerht|owt|eno").unwrap());

fn word_to_val(s: &str) -> Option<u32> {
    match s {
        "one" => Some(1),
        "two" => Some(2),
        "three" => Some(3),
        "four" => Some(4),
        "five" => Some(5),
        "six" => Some(6),
        "seven" => Some(7),
        "eight" => Some(8),
        "nine" => Some(9),
        "enin" => Some(9),
        "thgie" => Some(8),
        "neves" => Some(7),
        "xis" => Some(6),
        "evif" => Some(5),
        "ruof" => Some(4),
        "eerht" => Some(3),
        "owt" => Some(2),
        "eno" => Some(1),
        v => v.parse::<u32>().ok(),
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut sum1 = 0;
    let mut sum2 = 0;
    for line in reader.lines() {
        let line = line?;
        let rev = line.chars().rev().collect::<String>();
        {
            let first = DIGIT_ONLY_REGEX
                .find(&line)
                .and_then(|v| v.as_str().parse::<u32>().ok())
                .ok_or("Invalid digit")?;
            let last = DIGIT_ONLY_REGEX
                .find(&rev)
                .and_then(|v| v.as_str().parse::<u32>().ok())
                .ok_or("Invalid digit")?;
            sum1 += first * 10 + last;
        }
        {
            let first = DIGIT_REGEX
                .find(&line)
                .and_then(|v| word_to_val(v.as_str()))
                .ok_or("Invalid digit")?;
            let last = REV_DIGIT_REGEX
                .find(&rev)
                .and_then(|v| word_to_val(v.as_str()))
                .ok_or("Invalid digit")?;
            sum2 += first * 10 + last;
        }
    }

    println!("Part 1: {}", sum1);
    println!("Part 2: {}", sum2);
    Ok(())
}
